features/support/docker_gateway.rb

```diff
@@ -0,0 +1,38 @@
+# Ensure Docker container is completely stopped when Ruby exits.
+at_exit do
+  DockerGateway.new.stop
+end
+
+# Manages the Docker-based SSH server that is declared in docker-compose.yml.
+class DockerGateway
+  def initialize(log_proc=$stderr.method(:puts))
+    @log_proc = log_proc
+  end
+
+  def start
+    run_compose_command("up -d")
+  end
+
+  def stop
+    run_compose_command("down")
+  end
+
+  def run_shell_command(command)
+    run_compose_command("exec ssh_server /bin/bash -c #{command.shellescape}")
+  end
+
+  private
+
+  def run_compose_command(command)
+    log "[docker compose] #{command}"
+    stdout, stderr, status = Open3.capture3("docker compose #{command}")
```

Comment:
❓ Maybe we should stream the output while capturing it? `Open3.capture3` blocks until the process completes, so if the command is slow (e.g. downloading the docker base image), there won't be any output printed, and the test suite may appear "stuck" as a result.

Perhaps using a technique like this: https://mattbrictson.com/blog/run-shell-commands-in-ruby#streaming-output-while-capturing-it

Reply:
Thanks for the feedback, @G-Rath!

@mattbrictson I implemented something similar to your blog post as a new commit, but it's not quite seamless with the Cucumber logger. Two thing I noticed:

- Cucumber logger does not seem to flush messages as soon as they are received, and I suspect this is because it uses an STDOUT i/o whose `#sync == false`
- For reasons I don't grok, messages for image building (which can be numerous) and routine messages from docker-compose seem to be emitted differently, so `popen3` seemingly does not stream the latter (so we also continue the "capture then print" approach as before)

So I think this solves the radio silence issue, but symmetry with our Cucumber logger is still missing.

Reply:
@eric-eye ah, I see. Thanks for looking into that. My interpretation is that Cucumber does not print the `log` output until the test step finishes, which prevents us from streaming the output through that mechanism.

https://github.com/cucumber/cucumber-ruby/blob/main/lib/cucumber/formatter/pretty.rb#L116

I think your commit is a good compromise for now to fix the "radio silence" issue. Thanks!